File: java/account/AccountDelegate.java
/**
 * @file AccountDelegate.java
 * @author Ronald T. Fernandez
 * @version 1.0
 */
package account;

import java.util.Vector;

import corbaAccount.Account;
import corbaAccount.date;
import corbaAccount.Operation;

import account.OperationDelegate;
import connection.Connection;
import utils.Utils;

/**
 * This class implements the delegate for the user bank account
 * @see AccountImpl
 */
public class AccountDelegate {

	/**
	 * This is an instance of an AccountImpl, in which the class delegates
	 */
	private AccountImpl _instance;
	
	/**
	 * Constructor of the class
	 * @param name		Account owner name
	 * @param surname 	Account owner surname
	 * @param accountId	Account identifier
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountDelegate(String name, String surname, int accountId) throws Exception {
		this(name, surname, accountId, true);
	}

	/**
	 * Constructor of the class
	 * @param name		Account owner name
	 * @param surname 	Account owner surname
	 * @param accountId	Account identifier
	 * @param publish	Indicates whether the account needs to be added to the corba naming service
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountDelegate(String name, String surname, int accountId, boolean publish) throws Exception {
		_instance = new AccountImpl(name, surname, accountId);
		if (publish) {
			// Appends the accountId to the reference name in CORBA the naming service (accountId is assumed to be unique)
			Connection.getInstance().bindObjectToName(_instance._this(), "myContext", "Account"+_instance.accountId(), "Account");
		}
	}
	
	/**
	 * Constructor of the class
	 * @param name		Account owner name
	 * @param surname 	Account owner surname
	 * @param balance	Account initial balance
	 * @param accountId	Account identifier
	 * @throws Exception Throws an exception in case of any problem
	 */
	public AccountDelegate(String name, String surname, float balance, int accountId) throws Exception {
		_instance = new AccountImpl(name, surname, balance, accountId);
	}
	
	/**
	 * Obtain the account identifier
	 * @return accountId	Account id
	 */
	public int getAccountId() {
		return _instance.accountId();
	}
	
	/**
	 * Obtain the account owner name
	 * @return	Owner name
	 */
	public String getName() {
		return _instance.name();
	}
	
	/**
	 * Obtain the current acccount balance
	 * @return Up-to-date account balance
	 */
	public float getBalance() {
		return _instance.balance();
	}
	
	/**
	 * Show the account details
	 * @return String with a summary of the account
	 */
	public String toString() {
		return _instance.toString();
	}
	
	/**
	 * Get the account creation date
	 * @return Account creation date
	 * @throws Exception Throws an exception in case of any problem
	 */
	public DateDelegate getDate() throws Exception {
		date d = _instance.dateAccountCreated();
		return new DateDelegate(d.year(), d.month(), d.day());
	}
	
	/**
	 * Get the operations performed in the account
	 * @return List of operations
	 * @throws Exception Throws an exception in case of any problem
	 */
	public Vector<OperationDelegate> getOperations() throws Exception {
		Vector<OperationDelegate> v = new Vector<OperationDelegate>();
		// Convert the operations list in the CORBA format
		// to the returned vector type
		for(Operation op : _instance.accountOperations()) {
			v.addElement(new OperationDelegate(op.amount(), Utils.convertType(op.type()), op.operationId(), false));
		}
		return v;
	}
	
	/**
	 * Add an operation to the current account
	 * @param op	Operation to be added to the current account
	 */
	public void addOperation(OperationDelegate op) {
		_instance.addOperation(op.getCorbaInstance());
	}
	
	/**
	 * Get the CORBA instance for the account
	 * @return CORBA instance
	 */
	public Account getCorbaInstance() {
		return _instance.getCorbaInstance();
	}
		
}

File: cxx/account/AccountDelegate.cpp
/**
 * @file AccountDelegate.cpp
 * @author Ronald T. Fernandez
 * @version 1.0
 */

#include "../connection/Connection.h"

#include "AccountDelegate.h"

namespace account {

AccountDelegate::AccountDelegate(std::string name, std::string surname, int accountId, bool publish) {
	_instance = std::unique_ptr<AccountImpl>(new AccountImpl(name, surname, accountId));
	if (publish) {
		// Appends the accountId to the reference name in CORBA the naming service (accountId is assumed to be unique)
		connection::Connection::getInstance()->bindObjectToName(_instance->_this(), "myContext", "Account" + std::to_string(_instance->accountId()), "Account");
	}
}

AccountDelegate::AccountDelegate(std::string name, std::string surname, float balance, int accountId) {
	_instance = std::unique_ptr<AccountImpl>(new AccountImpl(name, surname, balance, accountId));
}

AccountDelegate::AccountDelegate(const AccountDelegate& ad) {
	_instance = std::unique_ptr<AccountImpl>(new AccountImpl(ad._instance->name(), ad._instance->surname(), ad._instance->balance(), ad._instance->accountId()));
}

AccountDelegate::~AccountDelegate() {
}

int AccountDelegate::getAccountId() const{
	return _instance->accountId();
}
std::string AccountDelegate::getName()  const{
	return _instance->name();
}

float AccountDelegate::getBalance() const {
	return _instance->balance();
}

std::string AccountDelegate::toString() const {
	return _instance->toString();
}
DateDelegate AccountDelegate::getDate() const {
	corbaAccount::date_ptr date = _instance->dateAccountCreated();
	return DateDelegate(date->year(), date->month(), date->day());
}
std::vector<OperationDelegate> AccountDelegate::getOperations() const {
	std::vector<OperationDelegate> operations;
	// Convert the operations list in the CORBA format
	// to the returned vector type
	corbaAccount::accountOperationsType* ops = _instance->accountOperations();
	size_t len = ops->length();
	for (size_t pos=0; pos<len; ++pos) {
		operations.push_back(OperationDelegate((*ops)[pos]->amount(), utils::Utils::convertType((*ops)[pos]->type()), (*ops)[pos]->operationId(), false));
	}
	return operations;
}
void AccountDelegate::addOperation(OperationDelegate op) {
	_instance->addOperation(op.getCorbaInstance());
}
corbaAccount::Account_ptr AccountDelegate::getCorbaInstance() const{
	return _instance->_this();
}

} /* namespace account */
